### PR TITLE
Implement first version of vgc::topology, vgc::workspace; Use Workspace in Canvas and app (#992)

### PR DESCRIPTION
**Pull request by PixelRick**
#992

### CODE REVIEW

**Comment by PixelRick** [comment]
Current bugs:
- crash related to elementsOutOfSync_ on update after doing some undo/redo.

File: libs/vgc/workspace/workspace.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+namespace detail {
+
+struct VacElementLists;
+
+} // namespace detail
+
+/// \class vgc::workspace::Workspace
+/// \brief Represents a workspace to manage, manipulate and visit a vector graphics scene.
+///
+class VGC_WORKSPACE_API Workspace : public core::Object {
```
Comment (dalboris) [inline libs/vgc/workspace/workspace.h:46]: Add comment: "This implementation is still preliminary and the API may change significantly"

File: libs/vgc/workspace/workspace.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    VGC_SLOT(onVacNodeAboutToBeRemoved, onVacNodeAboutToBeRemoved_);
+
+protected:
+    virtual void onDocumentDiff_(const dom::Diff& diff);
+    virtual void onVacDiff_(const topology::VacDiff& diff);
+
+private:
+    static std::unordered_map<core::StringId, ElementCreator>& elementCreators();
+
+    VacElement* vgcElement_;
```
Comment (dalboris) [inline libs/vgc/workspace/workspace.h:116]: Add comment: "this is the root element"

File: libs/vgc/workspace/workspace.cpp [hunk 3]
```diff
@@ -0,0 +1,813 @@
...
+    core::Array<Element*> groups;
+    core::Array<Element*> keyVertices;
+    core::Array<Element*> keyEdges;
+    core::Array<Element*> keyFaces;
+    core::Array<Element*> inbetweenVertices;
+    core::Array<Element*> inbetweenEdges;
+    core::Array<Element*> inbetweenFaces;
+};
+
+class ScopedRecursiveBoolGuard {
```
Comment (dalboris) [inline libs/vgc/workspace/workspace.cpp:44]: Add comment explaining the purpose of this.

**Review by dalboris — CHANGES_REQUESTED** [review]

**Review by dalboris — APPROVED** [review]

**Comment by PixelRick** [comment]
Related issue: #1107

**Pull request merged**